src/uct/ib/dc/dc_mlx5.c

```diff
@@ -483,36 +549,343 @@ static void uct_dc_mlx5_iface_cleanup_dcis(uct_dc_mlx5_iface_t *iface)
     }
 }
 
+static ucs_status_t uct_dc_mlx5_get_cmd_qp(uct_rc_mlx5_iface_common_t *rc_iface)
+{
+#if HAVE_DECL_MLX5DV_CONTEXT_FLAGS_DEVX
+    uct_dc_mlx5_iface_t *iface = ucs_derived_of(rc_iface, uct_dc_mlx5_iface_t);
+    uct_ib_md_t *md = uct_ib_iface_md(&iface->super.super.super);
+    uct_ib_device_t *dev = &md->dev;
+    uint32_t in[UCT_IB_MLX5DV_ST_SZ_DW(create_qp_in)] = {};
+    uint32_t out[UCT_IB_MLX5DV_ST_SZ_DW(create_qp_out)] = {};
+    uint32_t in_2init[UCT_IB_MLX5DV_ST_SZ_DW(rst2init_qp_in)] = {};
+    uint32_t out_2init[UCT_IB_MLX5DV_ST_SZ_DW(rst2init_qp_out)] = {};
+    uint32_t in_2rtr[UCT_IB_MLX5DV_ST_SZ_DW(init2rtr_qp_in)] = {};
+    uint32_t out_2rtr[UCT_IB_MLX5DV_ST_SZ_DW(init2rtr_qp_out)] = {};
+    uint32_t in_2rts[UCT_IB_MLX5DV_ST_SZ_DW(rtr2rts_qp_in)] = {};
+    uint32_t out_2rts[UCT_IB_MLX5DV_ST_SZ_DW(rtr2rts_qp_out)] = {};
+    ucs_status_t status = UCS_ERR_NO_MEMORY;
+    struct mlx5dv_pd dvpd = {};
+    struct mlx5dv_cq dvcq = {};
+    struct mlx5dv_obj dv = {};
+    int max = ucs_roundup_pow2(iface->super.tm.cmd_qp_len);
+    int len = max * UCT_IB_MLX5_MAX_BB * UCT_IB_MLX5_WQE_SEG_SIZE;
+    struct ibv_port_attr *port_attr;
+    uint8_t port_num;
+    uint32_t db_id;
+    size_t db_off;
+    void *qpc;
+    int ret;
+
+    port_num  = dev->first_port;
+    port_attr = uct_ib_device_port_attr(dev, port_num);
+
+    ret = posix_memalign(&iface->super.tm.cmd_wq.super.qstart,
+                         ucs_get_page_size(), len);
+    if (ret) {
+        return status;
+    }
+    iface->super.tm.cmd_wq.super.qend = iface->super.tm.cmd_wq.super.qstart + len;
+    iface->rx.cmd_qp_mem = mlx5dv_devx_umem_reg(dev->ibv_context,
```

Comment:
need to export a function like uct_ib_mlx5_devx_create_qp() to be reused in the future. it does not have to support everything at this stage, but at least be separated.